File: src/utils/embeddings.ts
/**
 * Embedding generation utilities using @xenova/transformers
 */

import { pipeline, AutoTokenizer, AutoModel, env } from '@xenova/transformers';

// Disable local model cache path warnings
env.allowLocalModels = false;

/**
 * Mean pooling function for transformer outputs
 */
function meanPooling(modelOutput: any, attentionMask: any): number[] {
  const tokenEmbeddings = modelOutput.last_hidden_state;
  const inputMaskExpanded = attentionMask
    .unsqueeze(-1)
    .expand(tokenEmbeddings.dims)
    .to(tokenEmbeddings.dtype);

  const sum = tokenEmbeddings.mul(inputMaskExpanded).sum(1);
  const sumMask = inputMaskExpanded.sum(1);
  const clampedSumMask = sumMask.clamp(1e-9, null);

  return sum.div(clampedSumMask).tolist()[0];
}

/**
 * Embedding generator class with model caching
 */
class EmbeddingGenerator {
  private modelCache: Map<string, { tokenizer: any; model: any }> = new Map();

  /**
   * Generate an embedding for the given text using the specified HuggingFace model ID
   */
  async generateEmbedding(text: string, modelId: string): Promise<number[]> {
    // Load or retrieve cached model
    const { tokenizer, model } = await this.loadModel(modelId);

    // Tokenize the input
    const inputs = await tokenizer(text, {
      padding: true,
      truncation: true,
    });

    // Run the model
    const output = await model(inputs);

    // Apply mean pooling
    const embedding = meanPooling(output, inputs.attention_mask);

    // Normalize the embedding (for cosine similarity)
    const norm = Math.sqrt(embedding.reduce((sum, val) => sum + val * val, 0));
    const normalizedEmbedding = embedding.map(val => val / norm);

    return normalizedEmbedding;
  }

  /**
   * Load a model and tokenizer, using cache if available
   */
  private async loadModel(huggingFaceModel: string): Promise<{ tokenizer: any; model: any }> {
    if (this.modelCache.has(huggingFaceModel)) {
      return this.modelCache.get(huggingFaceModel)!;
    }

    console.error(`Loading model: ${huggingFaceModel}...`);

    const tokenizer = await AutoTokenizer.from_pretrained(huggingFaceModel);
    const model = await AutoModel.from_pretrained(huggingFaceModel);

    this.modelCache.set(huggingFaceModel, { tokenizer, model });

    console.error(`Model loaded successfully.`);

    return { tokenizer, model };
  }

  /**
   * Clear the model cache
   */
  clearCache(): void {
    this.modelCache.clear();
  }
}

// Singleton instance
export const embeddingGenerator = new EmbeddingGenerator();

File: src/tpuff/utils/embeddings.py
"""Embedding generation utilities using sentence-transformers."""

import sys
from typing import Any

from rich.console import Console

console = Console(stderr=True)


class EmbeddingGenerator:
    """Embedding generator class with model caching."""

    def __init__(self) -> None:
        self._model_cache: dict[str, Any] = {}

    def generate_embedding(self, text: str, model_id: str) -> list[float]:
        """Generate an embedding for the given text using the specified model.

        Args:
            text: The text to embed.
            model_id: HuggingFace model ID (e.g., 'sentence-transformers/all-MiniLM-L6-v2').

        Returns:
            List of floats representing the embedding vector.
        """
        model = self._load_model(model_id)
        embedding = model.encode(text, normalize_embeddings=True)
        return embedding.tolist()

    def _load_model(self, model_id: str) -> Any:
        """Load a model, using cache if available.

        Args:
            model_id: HuggingFace model ID.

        Returns:
            Loaded SentenceTransformer model.
        """
        if model_id in self._model_cache:
            return self._model_cache[model_id]

        console.print(f"Loading model: {model_id}...", style="dim")

        try:
            from sentence_transformers import SentenceTransformer
            model = SentenceTransformer(model_id, trust_remote_code=True)
        except Exception as e:
            console.print(f"[red]Error loading model {model_id}: {e}[/red]")
            sys.exit(1)

        self._model_cache[model_id] = model
        console.print("Model loaded successfully.", style="dim")

        return model

    def clear_cache(self) -> None:
        """Clear the model cache."""
        self._model_cache.clear()


# Singleton instance
embedding_generator = EmbeddingGenerator()
